File: InterviewPreparationKit/LuckBalance.py
#!/bin/python3
# https://www.hackerrank.com/challenges/luck-balance/problem
import math
import os
import random
import re
import sys

# Complete the luckBalance function below.
def luckBalance(k, contests):
    balance = 0
    important = []

    for i in range(len(contests)):
        if contests[i][1] == 0:
            balance += contests[i][0]
        else:
            important.append(contests[i][0])

    important.sort()

    if len(important) > k:
        for j in range(len(important) - k, len(important)):
            balance += important[j]

        for m in range(len(important) - k):
            balance -= important[m]
    else:
        for p in range(len(important)):
            balance += important[p]

    return balance

if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    nk = input().split()

    n = int(nk[0])

    k = int(nk[1])

    contests = []

    for _ in range(n):
        contests.append(list(map(int, input().rstrip().split())))

    result = luckBalance(k, contests)

    fptr.write(str(result) + '\n')

    fptr.close()

File: Interview Preparation Kit/Greedy Algorithms/LuckBalance.cpp
// https://www.hackerrank.com/challenges/luck-balance/problem
#include <bits/stdc++.h>

using namespace std;

vector<string> split_string(string);

// Complete the luckBalance function below.
int luckBalance(int k, vector<vector<int>> contests) {
    int balance = 0;
    vector<int> important;

    int contests_rows = (int)contests.size();
    for (int i = 0; i < contests_rows; i++) {
        if (contests[i][1] == 0) {
            balance += contests[i][0];
        }
        else {
            important.insert(important.end(), contests[i][0]);
        }
    }

    sort(important.begin(), important.end());

    int important_size = (int)important.size();
    if (important_size > k) {
        int midpoint = important_size - k;
        for (int j = midpoint; j < important_size; j++) {
            balance += important[j];
        }

        for (int m = 0; m < midpoint; m++) {
            balance -= important[m];
        }
    }
    else {
        for (int p = 0; p < important_size; p++) {
            balance += important[p];
        }
    }

    return balance;
}

int main()
{
    ofstream fout(getenv("OUTPUT_PATH"));

    string nk_temp;
    getline(cin, nk_temp);

    vector<string> nk = split_string(nk_temp);

    int n = stoi(nk[0]);

    int k = stoi(nk[1]);

    vector<vector<int>> contests(n);
    for (int i = 0; i < n; i++) {
        contests[i].resize(2);

        for (int j = 0; j < 2; j++) {
            cin >> contests[i][j];
        }

        cin.ignore(numeric_limits<streamsize>::max(), '\n');
    }

    int result = luckBalance(k, contests);

    fout << result << "\n";

    fout.close();

    return 0;
}

vector<string> split_string(string input_string) {
    string::iterator new_end = unique(input_string.begin(), input_string.end(), [] (const char &x, const char &y) {
        return x == y and x == ' ';
    });

    input_string.erase(new_end, input_string.end());

    while (input_string[input_string.length() - 1] == ' ') {
        input_string.pop_back();
    }

    vector<string> splits;
    char delimiter = ' ';

    size_t i = 0;
    size_t pos = input_string.find(delimiter);

    while (pos != string::npos) {
        splits.push_back(input_string.substr(i, pos - i));

        i = pos + 1;
        pos = input_string.find(delimiter, i);
    }

    splits.push_back(input_string.substr(i, min(pos, input_string.length()) - i + 1));

    return splits;
}
